homeassistant/components/mqtt/__init__.py

```diff
@@ -939,10 +940,10 @@ def _mqtt_on_callback(self, _mqttc, _userdata, mid, _granted_qos=None) -> None:
         self.hass.add_job(self._mqtt_handle_mid, mid)
 
     async def _mqtt_handle_mid(self, mid) -> None:
-        if mid in self._pending_operations:
-            self._pending_operations[mid].set()
-        else:
-            _LOGGER.warning("Unknown mid %d", mid)
+        async with self._pending_operations_lock:
```

Comment:
I don't think this is a solution. The problem is that the mid is handled before the publish returns from the executor. So we need to use the existing paho lock.

Reply:
The point of the existing paho lock is to not have multiple threads make calls to the paho client. That's not an issue here so we don't need to use that lock.